File: shared/answerNormalize.ts
/**
 * Shared answer normalization for Web (+ mirrored in Flutter / smoke).
 *
 * Rules:
 * - null/undefined -> ""
 * - Unicode NFC (never NFD + strip marks)
 * - trim + collapse whitespace
 * - case fold only when caseInsensitive === true
 * - strip punctuation only when ignorePunctuation === true
 * - never remove accents / non-Latin / combining marks for answer checking
 */

export type AllowedScript =
  | "any"
  | "latin"
  | "kana"
  | "kanji"
  | "hangul"
  | "hanzi"
  | "arabic"
  | "thai"
  | "devanagari"
  | "cyrillic";

export interface NormalizeAnswerOptions {
  caseInsensitive?: boolean;
  ignorePunctuation?: boolean;
  /** Reserved for future input filters; default is "any" (no filtering). */
  allowedScript?: AllowedScript;
}

const PUNCTUATION_RE = /[.!?¡¿。、「」""''„«»]/gu;

export function normalizeAnswer(
  value: string | null | undefined,
  options: NormalizeAnswerOptions = {},
): string {
  if (value == null) return "";
  let text = String(value).normalize("NFC");
  text = text.trim().replace(/\s+/gu, " ");
  if (options.ignorePunctuation === true) {
    text = text.replace(PUNCTUATION_RE, "");
  }
  if (options.caseInsensitive === true) {
    text = text.toLocaleLowerCase();
  }
  return text;
}

export function answersMatch(
  submitted: string | null | undefined,
  expected: string | null | undefined,
  options: NormalizeAnswerOptions = {},
): boolean {
  return normalizeAnswer(submitted, options) === normalizeAnswer(expected, options);
}

File: mobile/novalang_flutter/lib/core/utils/answer_normalize.dart
/// Shared answer normalization (mirrors shared/answerNormalize.ts).
///
/// - null -> ""
/// - Preserve Unicode (no accent stripping, no non-Latin removal)
/// - Trim + collapse whitespace
/// - Case fold / punctuation strip only when explicitly requested
library;

class NormalizeAnswerOptions {
  const NormalizeAnswerOptions({
    this.caseInsensitive = false,
    this.ignorePunctuation = false,
    this.allowedScript = 'any',
  });

  final bool caseInsensitive;
  final bool ignorePunctuation;

  /// Reserved; default "any" means no script filtering on answers.
  final String allowedScript;
}

final RegExp _whitespaceCollapse = RegExp(r'\s+');
final RegExp _punctuation = RegExp(r'[.!?¡¿。、「」""''„«»]');

String normalizeAnswer(
  String? value, {
  NormalizeAnswerOptions options = const NormalizeAnswerOptions(),
}) {
  if (value == null) return '';
  // Keep all Unicode code points. Do not NFD-decompose or strip marks.
  // Full ICU NFC is not built into Dart; composed IME input is preserved as-is.
  var text = value.trim().replaceAll(_whitespaceCollapse, ' ');
  if (options.ignorePunctuation) {
    text = text.replaceAll(_punctuation, '');
  }
  if (options.caseInsensitive) {
    text = text.toLowerCase();
  }
  return text;
}

bool answersMatch(
  String? submitted,
  String? expected, {
  NormalizeAnswerOptions options = const NormalizeAnswerOptions(),
}) =>
    normalizeAnswer(submitted, options: options) ==
    normalizeAnswer(expected, options: options);
